src/MobileRedirectManager.php

```diff
@@ -0,0 +1,225 @@
+
+		if ( empty( $current_user_agent ) ) {
+			return false;
+		}
+
+		$is_mobile   = false;
+		$user_agents = self::get_user_agents();
+
+		foreach ( $user_agents as $user_agent ) {
+			if ( false !== strpos( $current_user_agent, $user_agent ) ) {
+				$is_mobile = true;
+				break;
+			}
+		}
+
+		return $is_mobile;
+	}
+
+	/**
+	 * Determine if mobile redirection should be done via JavaScript.
+	 *
+	 * @return bool True if mobile redirection should be done, false otherwise.
+	 */
+	public static function should_redirect_via_js() {
+		/**
+		 * Filters whether mobile redirection should be done via JavaScript. If false, a server-side solution will be used instead.
+		 *
+		 * @since 1.6
+		 *
+		 * @param bool $should_redirect_via_js Whether JS redirection should be used.
+		 */
+		return (bool) apply_filters( 'amp_redirect_via_js', true );
+	}
+
+	/**
+	 * Get a list of user agents to use for comparison against the user agent from the current request.
+	 *
+	 * @return string[] An array of user agents.
+	 */
+	public static function get_user_agents() {
+		// Default list compiled from the user agents listed in `wp_is_mobile()`.
+		$default_user_agents = [
+			'Mobile',
+			'Android',
+			'Silk/',
+			'Kindle',
+			'BlackBerry',
+			'Opera Mini',
+			'Opera Mobi',
+		];
+
+		/**
+		 * Filters the list of user agents used to determine if the user agent from the current request is a mobile one.
+		 *
+		 * @since 1.6
+		 *
+		 * @param string[] $user_agents List of user agents.
+		 */
+		return apply_filters( 'amp_mobile_user_agents', $default_user_agents );
+	}
+
+	/**
+	 * Determine if mobile redirection is disabled for the browser session.
+	 *
+	 * @return bool True if disabled, false otherwise.
+	 */
+	public static function redirection_disabled() {
+		return isset( $_COOKIE[ self::DISABLED_COOKIE_NAME ] ) && '1' === $_COOKIE[ self::DISABLED_COOKIE_NAME ];
+	}
+
+	/**
+	 * Output the mobile redirection Javascript code.
+	 */
+	public static function add_mobile_redirect_script() {
+		?>
+		<script>
+			(function ( ampSlug, disabledCookieName, userAgents ) {
+				const currentUserAgent = navigator.userAgent;
+				const isMobile = userAgents.some( ( userAgent ) => currentUserAgent.includes( userAgent ) );
+
+				if ( isMobile ) {
+					document.addEventListener( 'DOMContentLoaded', () => {
+						// Show the mobile version switcher link once the DOM has loaded.
+						const siteVersionSwitcher = document.getElementById( 'site-version-switcher' );
+						if ( siteVersionSwitcher ) {
+							siteVersionSwitcher.hidden = false;
+						}
+
+						// Disable mobile redirection if the user opts for the mobile version.
+						const versionSwitchLink = document.getElementById( 'version-switch-link' );
+						if ( versionSwitchLink ) {
+							versionSwitchLink.addEventListener( 'click', () => {
+								document.cookie = `${disabledCookieName}=;path=/;samesite=strict${ 'https:' === location.protocol ? ';secure' : '' }`;
```

Comment:
The JS used here needs to be IE11-compatible, so unfortunately no backticks or arrow functions.

Reply:
Done in 709ab6b687fde4f5c796309530b4bd78394b0c57.